File: Src/Icm.Core/Configuration/TypedDictionarySectionHandler.vb
Namespace Icm.Configuration
    Public Class TypedDictionarySectionHandler
        Implements System.Configuration.IConfigurationSectionHandler

        Public Function Create(ByVal parent As Object, ByVal configContext As Object, ByVal section As System.Xml.XmlNode) As Object Implements System.Configuration.IConfigurationSectionHandler.Create

            Dim ht As New Generic.Dictionary(Of String, TypedValue)
            Dim tv As TypedValue
            Dim t As Type
            For Each child As Xml.XmlNode In section.ChildNodes
                Select Case child.Name
                    Case "add"
                        If child.Attributes("type") Is Nothing Then
                            tv = New TypedValue(GetType(String), child.Attributes("value").Value)
                        Else
                            t = Type.GetType("System." & child.Attributes("type").Value)
                            tv = New TypedValue(t, Convert.ChangeType(child.Attributes("value").Value, Type.GetTypeCode(t)))
                        End If
                        ht.Add(child.Attributes("key").Value, tv)
                    Case "#comment"
                        ' Ignore comments
                    Case Else
                        Throw New Exception("<" & child.Name & ">: Not valid in a hashtable")
                End Select
            Next

            Return ht
        End Function

    End Class
End Namespace

File: Src/Icm.Core/Configuration/TypedDictionarySectionHandler.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Xml;

namespace Icm.Configuration
{
	public class TypedDictionarySectionHandler : System.Configuration.IConfigurationSectionHandler
	{

		public object Create(object parent, object configContext, System.Xml.XmlNode section)
		{
            
			var ht = new Dictionary<string, TypedValue>();
		    foreach (XmlNode child in section.ChildNodes) {
				switch (child.Name) {
					case "add":
				        TypedValue tv;
				        if (child.Attributes["type"] == null) {
							tv = new TypedValue(typeof(string), child.Attributes["value"].Value);
						} else
				        {
				            var t = Type.GetType("System." + child.Attributes["type"].Value);
				            tv = new TypedValue(t, Convert.ChangeType(child.Attributes["value"].Value, Type.GetTypeCode(t), CultureInfo.InvariantCulture));
				        }
				        ht.Add(child.Attributes["key"].Value, tv);
						break;
					case "#comment":
						break;
					// Ignore comments
					default:
						throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, "<{0}>: Not valid in a hashtable", child.Name));
				}
			}

			return ht;
		}

	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
